### PR TITLE
Reduce allocations for CreateDirectory

### PR DESCRIPTION
**Pull request by adamsitnik**
I wanted to see what would it take to have a sys-call that accepts a `ROS<char>` instead of `string` and after that I've discovered the existence of `ValueListBuilder<int>` which allowed me to get rid of `List<int>` allocation.

```ini
BenchmarkDotNet=v0.13.1.1616-nightly, OS=ubuntu 18.04
Intel Xeon CPU E5-1650 v4 3.60GHz, 1 CPU, 12 logical and 6 physical cores
.NET SDK=7.0.100-alpha.1.21566.20
```
|                         Method |       Toolchain | depth |       Mean | Ratio | Allocated |
|------------------------------- |---------------- |------ |-----------:|------:|----------:|
| RecursiveCreateDeleteDirectory | /before/corerun |    10 |   314.9 us |  1.00 |      8 KB |
| RecursiveCreateDeleteDirectory |  /after/corerun |    10 |   296.1 us |  0.95 |      7 KB |
|                                |                 |       |            |       |           |
| RecursiveCreateDeleteDirectory | /before/corerun |   100 | 3,565.2 us |  1.00 |    140 KB |
| RecursiveCreateDeleteDirectory |  /after/corerun |   100 | 3,559.1 us |  1.00 |    113 KB |

### CODE REVIEW

File: src/libraries/Common/src/Interop/Unix/System.Native/Interop.MkDir.cs
Comment (stephentoub) [inline src/libraries/Common/src/Interop/Unix/System.Native/Interop.MkDir.cs:13]: Doesn't GeneratedDllImport support spans directly?  Or maybe that hasn't been implemented yet? (@AaronRobinsonMSFT ?)

File: src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs:313]:
```suggestion
                using ValueListBuilder<int> stackDir = new(stackalloc int[32]); // 32 arbitrarily chosen
                CreateParentsAndDirectory(fullPath, stackDir);
```
?

File: src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs:321]:
```suggestion
        private static void CreateParentsAndDirectory(string fullPath, ref ValueListBuilder<int> stackDir)
```

**Review by stephentoub — APPROVED** [review]
Other than passing the VLB by ref (which needs to be done), LGTM.

File: src/libraries/Common/src/Interop/Unix/System.Native/Interop.MkDir.cs
Comment (carlossanlop) [inline src/libraries/Common/src/Interop/Unix/System.Native/Interop.MkDir.cs:11]: So since you're now passing a `ValueUtf8Converted`, this means the default value is `CharSet.Unicode`?

File: src/libraries/Common/src/Interop/Unix/System.Native/Interop.MkDir.cs
Comment (AaronRobinsonMSFT) [reply src/libraries/Common/src/Interop/Unix/System.Native/Interop.MkDir.cs:13]: Yes, we added testing support for `Span<T>`. It requires specifying the [`SpanMarshaller<T>`](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Runtime.InteropServices/tests/Ancillary.Interop/SpanMarshallers.cs) via a `MarshalUsing` attribute though. It is in the `Ancillary.Interop` assembly which is for our proposed public APIs the platform will provide.

More details on the marshaller are [here](https://github.com/dotnet/runtime/blob/main/docs/design/libraries/DllImportGenerator/SpanMarshallers.md).

/cc @jkoritzinsky @elinor-fung

Comment (adamsitnik) [reply src/libraries/Common/src/Interop/Unix/System.Native/Interop.MkDir.cs:13]: This is what I got when I've tried to use `ROS<char>`:

```log
error DLLIMPORTGEN001: The type 'System.ReadOnlySpan<char>' is not supported by source-generated P/Invokes. The generated source will not handle marshalling of parameter 'path'. [/home/adam/projects/runtime/src/coreclr/System.Private.CoreLib/System.Private.CoreLib.csproj]
```

It would be great if the error mentioned `SpanMarshaller<T>`

**Review by carlossanlop — COMMENTED** [review]
FYI @tmds has a PR open that's touching directory deletion APIs. 

Maybe the deletion APIs could benefit from a similar allocation reduction?

https://github.com/dotnet/runtime/pull/59520

**Review by carlossanlop — APPROVED** [review]
Oops, I meant to approve it, not comment.

LGTM (pending hearing from @AaronRobinsonMSFT in Stephen's question).

File: src/libraries/Common/src/Interop/Unix/System.Native/Interop.MkDir.cs
Comment (adamsitnik) [reply src/libraries/Common/src/Interop/Unix/System.Native/Interop.MkDir.cs:13]: @AaronRobinsonMSFT do I understand correctly that [`SpanMarshaller<T>`](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Runtime.InteropServices/tests/Ancillary.Interop/SpanMarshallers.cs) is currently in the test assembly and I can't easily use it in System.Private.CoreLib?

File: src/libraries/Common/src/Interop/Unix/System.Native/Interop.MkDir.cs
Comment (adamsitnik) [reply src/libraries/Common/src/Interop/Unix/System.Native/Interop.MkDir.cs:11]: iirc Windows sys-calls require Unicode while Unix Utf8

File: src/libraries/Common/src/Interop/Unix/System.Native/Interop.MkDir.cs
Comment (AaronRobinsonMSFT) [reply src/libraries/Common/src/Interop/Unix/System.Native/Interop.MkDir.cs:13]: It is in a proposed API assembly that will hopefully become part of the official API surface. One could copy it over and use right now&mdash;see [`ArrayMarshaller`](https://github.com/dotnet/runtime/blob/main/src/libraries/Common/src/System/Runtime/InteropServices/ArrayMarshaller.cs)&mdash;but ideally the source generator would just use it and not require user intervention.

If you want to consume it now then it can be done just requires some effort. Do we think the current scenario is hot enough to warrant the effort?

Comment (adamsitnik) [reply src/libraries/Common/src/Interop/Unix/System.Native/Interop.MkDir.cs:13]: @AaronRobinsonMSFT It can wait until it becomes part of the official API surface, thank you!

BTW I love the source-generated based marshalling as I finally know what is happening behind the scenes ;)

File: src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs
Comment (tmds) [inline src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs:304]: I'd keep the comment inside the `CreateParentsAndDirectory` method, as it explains what the method implementation is doing.

Can't you put the `ValueListBuilder<int>` in the method? And use: `using`?

File: src/libraries/Common/src/Interop/Unix/System.Native/Interop.MkDir.cs
Comment (tmds) [inline src/libraries/Common/src/Interop/Unix/System.Native/Interop.MkDir.cs:17]: same suggestion: use `using`.

File: src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs
Comment (adamsitnik) [reply src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs]: Great point! #resolved

**Review by tmds — APPROVED** [review]

**Comment by tmds** [comment]
I learned a few new tricks from this PR. Thanks @adamsitnik!

**Pull request merged**
